src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,183 @@
+    function scheddec<This, Ctx extends WorkflowContext, Return>(
+        target: object,
+        propertyKey: string,
+        inDescriptor: TypedPropertyDescriptor<(this: This, ctx: Ctx, ...args: ScheduledArgs) => Promise<Return>>
+    ) {
+        const { descriptor, registration } = registerAndWrapFunction(target, propertyKey, inDescriptor);
+        const schedRegistration = registration as unknown as SchedulerRegistration<This, ScheduledArgs, Return>;
+        schedRegistration.schedulerConfig = schedulerConfig;
+
+        return descriptor;
+    }
+    return scheddec;
+}
+
+///////////////////////////
+// Scheduler Management
+///////////////////////////
+
+export class DBOSScheduler{
+    constructor(readonly dbosExec: DBOSExecutor) {}
+
+    schedLoops: DetachableLoop[] = [];
+    schedTasks: Promise<void> [] = [];
+  
+    initScheduler() {
+        for (const registeredOperation of this.dbosExec.registeredOperations) {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                const loop = new DetachableLoop(this.dbosExec, ro.schedulerConfig.crontab ?? '* * * * *', ro);
+                this.schedLoops.push(loop);
+                this.schedTasks.push(loop.startLoop());
+            }
+        }
+    }
+  
+    async destroyScheduler() {
+        for (const l of this.schedLoops) {
+            await l.stopLoop();
+        }
+        this.schedLoops = [];
+        await Promise.all(this.schedTasks);
+        this.schedTasks = [];
+    }
+  
+    logRegisteredSchedulerEndpoints() {
+        const logger = this.dbosExec.logger;
+        logger.info("Scheduled endpoints:");
+        this.dbosExec.registeredOperations.forEach((registeredOperation) => {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab ?? '* * * * *'}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly mtd: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.mtd.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
```

Comment:
This is for handling crashes or a new node joining?

Reply:
This is the whole point exactly once.  If we don't have a durable record that it was kicked off, we do it again.  Any vm starting in a cluster with a history of executing the workflow will pick up this time and start.